core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -976,6 +975,239 @@ class ReplicaManager(val config: KafkaConfig,
+    producerId: Long,
+    producerEpoch: Short,
+    baseSequence: Int,
+    requestLocal: RequestLocal,
+    callback: (Errors, RequestLocal, VerificationGuard) => Unit
+  ): Unit = {
+    def generalizedCallback(preAppendErrors: Map[TopicPartition, Errors],
+                            newRequestLocal: RequestLocal,
+                            verificationGuards: Map[TopicPartition, VerificationGuard]): Unit = {
+      callback(
+        preAppendErrors.getOrElse(topicPartition, Errors.NONE),
+        newRequestLocal,
+        verificationGuards.getOrElse(topicPartition, VerificationGuard.SENTINEL))
+    }
+
+    maybeStartTransactionVerificationForPartitions(
+      Map(topicPartition -> baseSequence),
+      transactionalId,
+      producerId,
+      producerEpoch,
+      requestLocal,
+      generalizedCallback
+    )
+  }
+
+  def maybeStartTransactionVerificationForPartitions(
+    topicPartitionBatchInfo: Map[TopicPartition, Int],
+    transactionalId: String,
+    producerId: Long,
+    producerEpoch: Short,
+    requestLocal: RequestLocal,
+    callback: (Map[TopicPartition, Errors], RequestLocal, Map[TopicPartition, VerificationGuard]) => Unit
+  ): Unit = {
+    // Skip verification if the request is not transactional or transaction verification is disabled.
+    if (transactionalId == null ||
+      !config.transactionPartitionVerificationEnable
+      || addPartitionsToTxnManager.isEmpty
+    ) {
+      callback(Map.empty[TopicPartition, Errors], requestLocal, Map.empty[TopicPartition, VerificationGuard])
+      return
+    }
+
+    // Wrap the callback to be handled on an arbitrary request handler thread when transaction verification is complete.
+    val verificationGuards = mutable.Map[TopicPartition, VerificationGuard]()
+    val errors = mutable.Map[TopicPartition, Errors]()
+
+    topicPartitionBatchInfo.map { case (topicPartition, baseSequence) =>
+      val errorOrGuard = maybeStartTransactionVerificationForPartition(
+        topicPartition,
+        producerId,
+        producerEpoch,
+        baseSequence
+      )
+
+      errorOrGuard match {
+        case Left(error) => errors.put(topicPartition, error)
+        case Right(verificationGuard) => if (verificationGuard != VerificationGuard.SENTINEL)
+          verificationGuards.put(topicPartition, verificationGuard)
+      }
+    }
+
+    // All partitions are already verified
+    if (verificationGuards.isEmpty) {
+      callback(Map.empty[TopicPartition, Errors], requestLocal, Map.empty[TopicPartition, VerificationGuard])
+      return
+    }
+
+    def invokeCallback(
+      requestLocal: RequestLocal,
+      verificationErrors: Map[TopicPartition, Errors]
+    ): Unit = {
+      callback(errors ++ verificationErrors, requestLocal, verificationGuards.toMap)
+    }
+
+    // Wrap the callback to be handled on an arbitrary request handler thread when transaction verification is complete.
+    val verificationCallback = KafkaRequestHandler.wrapAsyncCallback(
```

Comment:
Just a comment for future (we can address it when we implement functionality for the new GC) -- in the new GC the callback gets immediately scheduled on the GC thread pool, so we should move wrapAsyncCallback responsibility to the caller (old GC would do it, but the new GC won't).

Reply:
Yeah -- I was thinking about this before too. I was also talking to @dajac about if we should move to the new gc threads.